File: lib/typr/words.ex
defmodule Typr.Words do
  @moduledoc """
  Word lists and test-text generation.

  Two vocabularies ship with `typr`: `english`, the most frequent words in the
  language, and `english_extended`, which adds longer and less common words.
  Both can be decorated with punctuation and numbers the way monkeytype does.
  """

  @english ~w(
    the be to of and a in that have it for not on with he as you do at this but
    his by from they we say her she or an will my one all would there their what
    so up out if about who get which go me when make can like time no just him
    know take people into year your good some could them see other than then now
    look only come its over think also back after use two how our work first well
    way even new want because any these give day most us man find here thing tell
    very still should much need right old too same life world own feel three high
    place small large next early young important few public bad able part hand eye
    woman child home water room mother area money story fact month night book word
    side kind head house friend father hour game line end member car city name team
    minute idea kid body face level door health person art war history party result
    change morning reason research girl guy moment air teacher force education foot
    boy age policy process music market sense nation plan college interest death
    course someone experience behind reach local sure president road table sport
    talk turn start might hard open walk white number group show run move live
    believe hold bring happen write sit stand lose pay meet include continue set
    learn lead understand watch follow stop create speak read spend grow drive
    break call try ask keep let begin seem help
  )

  @english_extended ~w(
    absolute achieve alongside amount analysis appear approach argument assume
    balance benefit beyond capital carefully category challenge character
    circumstance collection combine comfortable community complete concern
    condition confidence connection consider constant contain context
    contribute conversation critical culture current decision definitely
    describe design detail determine develop difference direction discover
    discussion distance distribute economy effective effort element encourage
    energy environment equipment establish evidence exactly example excellent
    exchange exercise expect explain express extremely feature figure finally
    financial function further generation generous global hardly however
    identify imagine immediate improve include increase indicate individual
    industry influence information initial instead institution intelligence
    introduce investment involve knowledge language largely leadership
    literature machine maintain majority management material measure mention
    method modern movement natural necessary negative neighborhood network
    normally obviously occasion opportunity organize original otherwise
    particular perform perhaps performance permanent personal perspective
    physical popular position possible potential practice prepare presence
    prevent previous private probably produce professional program property
    protect provide purpose quality quarter question quickly realize
    recognize recommend reduce reflect regular relationship relative remain
    remember remove replace represent require resource respond responsible
    restaurant returning revenue satisfy schedule science section security
    separate serious service several signal significant similar situation
    society software solution somewhere specific standard strategy structure
    successful sufficient suggest support suppose surface surround survive
    technology television temperature terrible therefore thousand throughout
    tomorrow tradition traffic transfer treatment tremendous typical
    understand university unusual valuable variety vehicle version violence
    vision volume welcome whatever whenever wherever whether whisper without
    wonderful writing yesterday
  )

  # Pangrams and plain proverbs, written for this project so the text stays
  # unencumbered. Quote mode draws from these.
  @sentences [
    "the quick brown fox jumps over the lazy dog while the sleepy cat watches from a sunny window",
    "pack my box with five dozen liquor jugs and ship it out before the harbor freezes over",
    "how vexingly quick daft zebras jump when a jackal wanders past the waterhole at dusk",
    "a journey of a thousand miles begins with a single step taken in the right direction",
    "measure twice and cut once because the board never grows back once you have trimmed it",
    "the best time to plant a tree was twenty years ago and the second best time is today",
    "practice does not make perfect but it does make permanent so practice the thing you want to keep",
    "a smooth sea never made a skilled sailor and calm weather teaches nothing about the storm",
    "speed comes from accuracy repeated often enough that your hands stop asking your eyes for help",
    "the person who moves a mountain begins by carrying away one small stone at a time",
    "we judge ourselves by what we mean to do and everyone else by what they actually did",
    "sharpen the axe before you swing it or you will spend the whole day making noise in the forest"
  ]

  @punctuation_marks [".", ".", ".", ",", ",", ";", ":", "!", "?"]

  @doc "Names of the available word lists."
  @spec list_names() :: [String.t()]
  def list_names, do: ["english", "english_extended"]

  @doc "The raw vocabulary behind a list name."
  @spec vocabulary(String.t()) :: [String.t()] | nil
  def vocabulary("english"), do: @english
  def vocabulary("english_extended"), do: @english_extended
  def vocabulary(_), do: nil

  @doc "A random sentence for quote mode, as a list of words."
  @spec quote_words() :: [String.t()]
  def quote_words do
    @sentences |> Enum.random() |> String.split()
  end

  @doc """
  Generates `count` words from `list`, applying the `:punctuation` and
  `:numbers` options.

  Consecutive duplicates are avoided so the text reads like prose rather than
  a stutter. When punctuation is on, the first word is capitalized and every
  sentence-ending mark capitalizes the word that follows it.
  """
  @spec generate(String.t(), pos_integer(), keyword()) :: [String.t()]
  def generate(list, count, opts \\ []) do
    vocab = vocabulary(list) || @english

    vocab
    |> random_words(count)
    |> decorate(opts)
  end

  defp random_words(vocab, count) do
    Enum.map_reduce(1..count, nil, fn _, previous ->
      word = draw(vocab, previous)
      {word, word}
    end)
    |> elem(0)
  end

  defp draw(vocab, previous) do
    case Enum.random(vocab) do
      ^previous -> draw(vocab, previous)
      word -> word
    end
  end

  defp decorate(words, opts) do
    numbers? = Keyword.get(opts, :numbers, false)
    punctuation? = Keyword.get(opts, :punctuation, false)

    words
    |> maybe(numbers?, &add_numbers/1)
    |> maybe(punctuation?, &add_punctuation/1)
  end

  defp maybe(words, true, fun), do: fun.(words)
  defp maybe(words, false, _fun), do: words

  defp add_numbers(words) do
    Enum.map(words, fn word ->
      if :rand.uniform() < 0.12 do
        digits = Enum.random(1..4)
        Integer.to_string(Enum.random(0..(Integer.pow(10, digits) - 1)))
      else
        word
      end
    end)
  end

  # Walks the list carrying "does the next word start a sentence?" so capitals
  # land where a reader would expect them.
  defp add_punctuation(words) do
    words
    |> Enum.map_reduce(true, fn word, sentence_start? ->
      {word, ends_sentence?} = punctuate(word)
      word = if sentence_start?, do: capitalize(word), else: word
      {word, ends_sentence?}
    end)
    |> elem(0)
  end

  defp punctuate(word) do
    roll = :rand.uniform()

    cond do
      roll < 0.06 -> {wrap(word), false}
      roll < 0.10 -> {word <> "'s", false}
      roll < 0.30 -> append_mark(word)
      true -> {word, false}
    end
  end

  defp append_mark(word) do
    mark = Enum.random(@punctuation_marks)
    {word <> mark, mark in [".", "!", "?"]}
  end

  defp wrap(word) do
    {open, close} = Enum.random([{"(", ")"}, {"\"", "\""}, {"'", "'"}, {"[", "]"}])
    open <> word <> close
  end

  # Capitalizes the first letter even when the word opens with a bracket or
  # quote, and leaves the rest of the word alone.
  defp capitalize(word) do
    case String.next_grapheme(word) do
      {opener, rest} when opener in ["(", "\"", "'", "["] -> opener <> upcase_first(rest)
      _ -> upcase_first(word)
    end
  end

  defp upcase_first(word) do
    case String.next_grapheme(word) do
      {first, rest} -> String.upcase(first) <> rest
      nil -> word
    end
  end
end

File: src/words.rs
//! Word lists and test-text generation.
//!
//! Two vocabularies ship with `typr`: `english`, the most frequent words in the
//! language, and `english_extended`, which adds longer and less common words.
//! Both can be decorated with punctuation and numbers the way monkeytype does.

use crate::rng::Rng;
use crate::words_data::{ENGLISH, ENGLISH_EXTENDED, SENTENCES};

const PUNCTUATION_MARKS: &[&str] = &[".", ".", ".", ",", ",", ";", ":", "!", "?"];
const BRACKETS: &[(&str, &str)] = &[("(", ")"), ("\"", "\""), ("'", "'"), ("[", "]")];
const SENTENCE_ENDINGS: &[&str] = &[".", "!", "?"];

/// Names of the available word lists.
pub fn list_names() -> &'static [&'static str] {
    &["english", "english_extended"]
}

/// The raw vocabulary behind a list name.
pub fn vocabulary(name: &str) -> Option<&'static [&'static str]> {
    match name {
        "english" => Some(ENGLISH),
        "english_extended" => Some(ENGLISH_EXTENDED),
        _ => None,
    }
}

/// A random sentence for quote mode, as a list of words.
pub fn quote_words(rng: &mut Rng) -> Vec<String> {
    rng.choose(SENTENCES)
        .split_whitespace()
        .map(str::to_string)
        .collect()
}

/// What to mix into otherwise plain text.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct Decoration {
    pub punctuation: bool,
    pub numbers: bool,
}

/// Generates `count` words from `list`, applying `decoration`.
///
/// Consecutive duplicates are avoided so the text reads like prose rather than
/// a stutter. When punctuation is on, the first word is capitalized and every
/// sentence-ending mark capitalizes the word that follows it.
pub fn generate(list: &str, count: usize, decoration: Decoration, rng: &mut Rng) -> Vec<String> {
    let vocab = vocabulary(list).unwrap_or(ENGLISH);
    let mut words = random_words(vocab, count, rng);

    if decoration.numbers {
        add_numbers(&mut words, rng);
    }

    if decoration.punctuation {
        add_punctuation(&mut words, rng);
    }

    words
}

fn random_words(vocab: &[&str], count: usize, rng: &mut Rng) -> Vec<String> {
    let mut words: Vec<String> = Vec::with_capacity(count);
    let mut previous: Option<&str> = None;

    for _ in 0..count {
        let word = draw(vocab, previous, rng);
        previous = Some(word);
        words.push(word.to_string());
    }

    words
}

fn draw<'a>(vocab: &'a [&str], previous: Option<&str>, rng: &mut Rng) -> &'a str {
    loop {
        let word = *rng.choose(vocab);
        if Some(word) != previous {
            return word;
        }
    }
}

fn add_numbers(words: &mut [String], rng: &mut Rng) {
    for word in words.iter_mut() {
        if rng.fraction() < 0.12 {
            let digits = rng.between(1, 4);
            let ceiling = 10u32.pow(digits) - 1;
            *word = rng.between(0, ceiling).to_string();
        }
    }
}

/// Walks the list carrying "does the next word start a sentence?" so capitals
/// land where a reader would expect them.
fn add_punctuation(words: &mut [String], rng: &mut Rng) {
    let mut sentence_start = true;

    for word in words.iter_mut() {
        let (punctuated, ends_sentence) = punctuate(word, rng);

        *word = if sentence_start {
            capitalize(&punctuated)
        } else {
            punctuated
        };

        sentence_start = ends_sentence;
    }
}

fn punctuate(word: &str, rng: &mut Rng) -> (String, bool) {
    let roll = rng.fraction();

    if roll < 0.06 {
        (wrap(word, rng), false)
    } else if roll < 0.10 {
        (format!("{word}'s"), false)
    } else if roll < 0.30 {
        append_mark(word, rng)
    } else {
        (word.to_string(), false)
    }
}

fn append_mark(word: &str, rng: &mut Rng) -> (String, bool) {
    let mark = *rng.choose(PUNCTUATION_MARKS);
    (format!("{word}{mark}"), SENTENCE_ENDINGS.contains(&mark))
}

fn wrap(word: &str, rng: &mut Rng) -> String {
    let (open, close) = *rng.choose(BRACKETS);
    format!("{open}{word}{close}")
}

/// Capitalizes the first letter even when the word opens with a bracket or
/// quote, and leaves the rest of the word alone.
fn capitalize(word: &str) -> String {
    let mut characters = word.chars();

    match characters.next() {
        None => String::new(),
        Some(opener) if matches!(opener, '(' | '"' | '\'' | '[') => {
            format!("{opener}{}", upcase_first(characters.as_str()))
        }
        Some(_) => upcase_first(word),
    }
}

fn upcase_first(word: &str) -> String {
    let mut characters = word.chars();

    match characters.next() {
        Some(first) => first.to_uppercase().collect::<String>() + characters.as_str(),
        None => String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    const PUNCTUATED: Decoration = Decoration {
        punctuation: true,
        numbers: false,
    };
    const NUMBERED: Decoration = Decoration {
        punctuation: false,
        numbers: true,
    };
    const BOTH: Decoration = Decoration {
        punctuation: true,
        numbers: true,
    };

    fn rng() -> Rng {
        Rng::seeded(20260831)
    }

    fn starts_a_sentence(word: &str) -> bool {
        let mut characters = word.chars();

        let first = match characters.next() {
            Some(character) => character,
            None => return false,
        };

        if matches!(first, '(' | '"' | '\'' | '[') {
            characters.next().is_some_and(|next| next.is_uppercase())
        } else {
            first.is_uppercase()
        }
    }

    #[test]
    fn produces_the_requested_number_of_words() {
        assert_eq!(
            generate("english", 25, Decoration::default(), &mut rng()).len(),
            25
        );
    }

    #[test]
    fn never_repeats_a_word_back_to_back() {
        let words = generate("english", 500, Decoration::default(), &mut rng());

        assert!(words.windows(2).all(|pair| pair[0] != pair[1]));
    }

    #[test]
    fn draws_only_from_the_requested_vocabulary() {
        let vocabulary: HashSet<&str> = vocabulary("english_extended")
            .unwrap()
            .iter()
            .copied()
            .collect();

        assert!(
            generate("english_extended", 100, Decoration::default(), &mut rng())
                .iter()
                .all(|word| vocabulary.contains(word.as_str()))
        );
    }

    #[test]
    fn falls_back_to_the_default_vocabulary_for_an_unknown_list() {
        assert_eq!(
            generate("klingon", 10, Decoration::default(), &mut rng()).len(),
            10
        );
    }

    #[test]
    fn capitalises_the_opening_word() {
        for seed in 0..20 {
            let words = generate("english", 10, PUNCTUATED, &mut Rng::seeded(seed));

            assert!(
                starts_a_sentence(&words[0]),
                "{:?} does not open a sentence",
                words[0]
            );
        }
    }

    #[test]
    fn adds_marks_that_plain_generation_never_produces() {
        let marks = ['.', ',', ';', ':', '!', '?'];

        let plain = generate("english", 300, Decoration::default(), &mut rng()).join(" ");
        let punctuated = generate("english", 300, PUNCTUATED, &mut rng()).join(" ");

        assert!(!plain.contains(marks));
        assert!(punctuated.contains(marks));
    }

    #[test]
    fn a_word_after_a_full_stop_starts_a_new_sentence() {
        let words = generate("english", 400, PUNCTUATED, &mut rng());

        for pair in words.windows(2) {
            let ends_sentence = pair[0].ends_with(['.', '!', '?']);

            if ends_sentence {
                assert!(
                    starts_a_sentence(&pair[1]),
                    "{:?} follows {:?} but is not capitalised",
                    pair[1],
                    pair[0]
                );
            }
        }
    }

    #[test]
    fn mixes_digits_in_when_asked_and_never_otherwise() {
        let with = generate("english", 400, NUMBERED, &mut rng()).join("");
        let without = generate("english", 400, Decoration::default(), &mut rng()).join("");

        assert!(with.chars().any(|character| character.is_ascii_digit()));
        assert!(!without.chars().any(|character| character.is_ascii_digit()));
    }

    #[test]
    fn quote_words_returns_a_sentence_as_separate_words() {
        let words = quote_words(&mut rng());

        assert!(words.len() > 5);
        assert!(words.iter().all(|word| !word.contains(' ')));
    }

    #[test]
    fn every_listed_name_resolves() {
        assert!(list_names().iter().all(|name| vocabulary(name).is_some()));
    }

    #[test]
    fn unknown_names_do_not() {
        assert!(vocabulary("nope").is_none());
    }

    #[test]
    fn lists_hold_no_duplicates_or_stray_whitespace() {
        for name in list_names() {
            let vocabulary = vocabulary(name).unwrap();
            let unique: HashSet<&&str> = vocabulary.iter().collect();

            assert_eq!(unique.len(), vocabulary.len(), "{name} has duplicates");
            assert!(vocabulary.iter().all(|word| word.trim() == *word));
            assert!(vocabulary.iter().all(|word| !word.is_empty()));
        }
    }

    #[test]
    fn the_same_seed_generates_the_same_test() {
        let first = generate("english", 50, BOTH, &mut Rng::seeded(99));
        let second = generate("english", 50, BOTH, &mut Rng::seeded(99));

        assert_eq!(first, second);
    }
}
